onnxruntime/core/providers/opencl/opencl_utils.h

```diff
@@ -0,0 +1,412 @@
+    ORT_ENFORCE(shape.NumDimensions() == 1);
+    //
+    return {1024, CeilDiv(shape[0], 4 * 1024)};
+  }
+
+  static Image2DDesc PackFromTensor2D(const TensorShape& shape) {
+    ORT_ENFORCE(shape.NumDimensions() == 2);
+    return {CeilDiv(shape[0], 4), shape[1]};
+  }
+
+  static Image2DDesc PackFromTensorNCHW(const TensorShape& shape) {
+    ORT_ENFORCE(shape.NumDimensions() == 4);
+    int64_t N = shape[0];
+    int64_t C = shape[1];
+    int64_t H = shape[2];
+    int64_t W = shape[3];
+    int64_t c = 4;
+    int64_t Cc = CeilDiv(C, c);
+    return {Cc * W, N * H};
+  }
+
+  // NCHWc is actually Tensor of shape N[C/c]HWc then packed as NH C/cWc
+  static Image2DDesc PackFromTensorNCHWc(const TensorShape& shape) {
+    ORT_ENFORCE(shape.NumDimensions() == 5);
+    int64_t N = shape[0];
+    int64_t Cc = shape[1];
+    int64_t H = shape[2];
+    int64_t W = shape[3];
+    int64_t c = shape[4];
+    ORT_ENFORCE(c == 4);
+    return {Cc * W, N * H};
+  }
+
+  static Image2DDesc PackFromConv2DWeight(const TensorShape& shape) {
+    ORT_ENFORCE(shape.NumDimensions() == 4);
+    int64_t C_o = shape[0];
+    int64_t C_i = shape[1];
+    int64_t K_h = shape[2];
+    int64_t K_w = shape[3];
+    return {C_i, CeilDiv(C_o, 4) * K_h * K_w};
+  }
+  static Image2DDesc PackFromWinogradTransform(const TensorShape& shape) {
+    ORT_ENFORCE(shape.NumDimensions() == 4);
+    int64_t C_o = shape[0];
+    int64_t C_i = shape[1];
+    //FIXME: asumme we only surpport window-size=4
+    int64_t K_h = 4;
+    int64_t K_w = 4;
+    return {CeilDiv(C_i, 4) * K_h, 16*CeilDiv(C_o, 4)};
+  }
+
+  static Image2DDesc PackFromDepthwiseConv2DWeight(const TensorShape& shape) {
+    ORT_ENFORCE(shape.NumDimensions() == 4);
+    int64_t C_o = shape[0];
+    int64_t C_i = shape[1];
+    int64_t K_h = shape[2];
+    int64_t K_w = shape[3];
+    return {K_h * K_w * C_i, CeilDiv(C_o, 4)};
+  }
+
+  auto Height() const {
+    return second;
+  }
+
+  auto Width() const {
+    return first;
+  }
+
+  size_t UHeight() const {
+    return static_cast<size_t>(second);
+  }
+
+  size_t UWidth() const {
+    return static_cast<size_t>(first);
+  }
+
+  NDRange AsNDRange() const {
+    return {UWidth(), UHeight()};
+  }
+
+  bool operator==(const onnxruntime::opencl::Image2DDesc& other) const {
+    return Width() == other.Width() && Height() == other.Height();
+  }
+};
+
+class KernelLauncher {
+ public:
+  explicit KernelLauncher(cl_kernel kernel) : kernel_{kernel}, index_{0}, err_{CL_SUCCESS} {}
+  cl_kernel Kernel() const { return kernel_; }
+
+#define SKIP_IF_ERRORED(expr) \
+  if (err_ == CL_SUCCESS) {   \
+    err_ = (expr);            \
+    err_index_ = index_;      \
+  }
+
+  /**
+   * @brief Set the dynamic local memory size. (aka., shared memory in CUDA).
+   *
+   * This function can be called multiple times.
+   */
+  template <typename T, typename E = std::is_convertible<T, size_t>>
+  KernelLauncher& setShmem(T num_bytes) {
```

Comment:
nit: naming conventions. SetShmem. same for other methods here.

Reply:
Done